logstash-core/src/main/java/org/logstash/secret/store/backend/JavaKeyStore.java

```diff
@@ -0,0 +1,227 @@
+
+                    PosixFileAttributeView attrs = Files.getFileAttributeView(keyStorePath, PosixFileAttributeView.class);
+                    if (attrs != null) {
+                        attrs.setPermissions(PosixFilePermissions.fromString("rw-rw----"));
+                    }
+                }
+            }
+        } catch (Exception e) {
+            throw new SecretStoreException("Error while construction the JavaKeyStore", e);
+        }
+    }
+
+    /**
+     * Converts bytes from ascii encoded text to a char[] and zero outs the original byte[]
+     *
+     * @param bytes the bytes from an ascii encoded text (note - no validation is done to ensure ascii encoding)
+     * @return the corresponding char[]
+     */
+    private char[] asciiBytesToChar(byte[] bytes) {
+        char[] chars = new char[bytes.length];
+        for (int i = 0; i < bytes.length; i++) {
+            chars[i] = (char) bytes[i];
+            bytes[i] = '\0';
+        }
+        return chars;
+    }
+
+    /**
+     * Converts characters from ascii encoded text to a byte[] and zero outs the original char[]
+     *
+     * @param chars the chars from an ascii encoded text (note - no validation is done to ensure ascii encoding)
+     * @return the corresponding byte[]
+     */
+    private byte[] asciiCharToBytes(char[] chars) {
+        byte[] bytes = new byte[chars.length];
+        for (int i = 0; i < chars.length; i++) {
+            bytes[i] = (byte) chars[i];
+            chars[i] = '\0';
+        }
+        return bytes;
+    }
+
+    /**
+     * Attempt to keep the secret data out of the heap.
+     *
+     * @param secret the secret to zero out
+     */
+    private void clearSecret(byte[] secret) {
+        if (secret != null) {
+            for (int i = 0; i < secret.length; ++i) {
+                secret[i] = '\0';
+            }
+            secret = null;
+        }
+    }
+
+    @Override
+    public Collection<SecretIdentifier> list() {
+        Set<SecretIdentifier> identifiers = new HashSet<>();
+        try {
+            readLock.lock();
+            Enumeration<String> aliases = keyStore.aliases();
+            while (aliases.hasMoreElements()) {
+                String alias = aliases.nextElement();
+                identifiers.add(SecretIdentifier.fromExternalForm(alias));
+            }
+        } catch (KeyStoreException e) {
+            e.printStackTrace();
+        } finally {
+            readLock.unlock();
+        }
+        return identifiers;
+    }
+
+    @Override
+    public void persistSecret(SecretIdentifier identifier, byte[] secret) {
+        try {
+            writeLock.lock();
+            SecretKeyFactory factory = SecretKeyFactory.getInstance("PBE");
+            //PBEKey requires an ascii password, so base64 encode it
+            byte[] base64 = Base64.getEncoder().encode(secret);
+            PBEKeySpec passwordBasedKeySpec = new PBEKeySpec(asciiBytesToChar(base64));
```

Comment:
Thoughts on adding types in the future? 

In this PR, there are only byte[] types, but it's possible some future where we have a way to safely persist other types, or at least an abstraction to persist other types. It's possible we don't' need this at all, but I always think about this when considering serialization.

Related: Elasticsearch has a slight wrapper, SecureString, which wraps a char[] and is both CharSequence and Closeable -- we may consider using. Thoughts?

Reply:
For now, I think byte[] suites us for the API contract, and want to code against the API so that we can implement other backends, or allow 3rd parties to bring their own secret store.  

As an implementation detail, all of the data, regardless of upstream type is re-encoded as base64 char[] to ensure ASCII compatibility with the underlying library. Any additional types supported would be essentially be helper methods to get the serialized form (e.g. byte[]). 

I would think that the functions/helpers to get to the serialized form should be the responsibility of the given type, not the secret store.